lib/ansible/modules/crypto/openssh_keypair.py

```diff
@@ -0,0 +1,315 @@
+                                      'Attempting to use bit lengths other than these three values for '
+                                      'ECDSA keys will cause this module to fail. '))
+        if self.type == 'ed25519':
+            self.size = 256
+
+    def generate(self, module):
+        # generate a keypair
+        if not self.check(module, perms_required=False) or self.force:
+            args = [
+                module.get_bin_path('ssh-keygen', True),
+                '-q',
+                '-N', '',
+                '-b', str(self.size),
+                '-t', self.type,
+                '-f', self.path,
+            ]
+
+            if self.comment:
+                args.extend(['-C', self.comment])
+            else:
+                args.extend(['-C', ""])
+
+            try:
+                self.changed = True
+                module.run_command(args)
+                proc = module.run_command([module.get_bin_path('ssh-keygen', True), '-lf', self.path])
+                self.fingerprint = proc[1].split()
+            except Exception as e:
+                self.remove()
+                module.fail_json(msg="%s" % to_native(e))
+
+        file_args = module.load_file_common_arguments(module.params)
+        if module.set_fs_attributes_if_different(file_args, False):
+            self.changed = True
+
+    def check(self, module, perms_required=True):
+
+        # check if the key is correct
+        def _check_state():
+            return os.path.exists(self.path)
+
+        if _check_state():
+            proc = module.run_command([module.get_bin_path('ssh-keygen', True), '-lf', self.path])
+            fingerprint = proc[1].split()
+            keysize = int(fingerprint[0])
+            keytype = fingerprint[-1][1:-1].lower()
+        else:
+            return False
+
+        def _check_perms(module):
+            file_args = module.load_file_common_arguments(module.params)
+            return not module.set_fs_attributes_if_different(file_args, False)
+
+        def _check_type():
+            return self.type == keytype
+
+        def _check_size():
+            return self.size == keysize
+
+        self.fingerprint = fingerprint
+
+        if not perms_required:
+            return _check_state() and _check_type() and _check_size()
+
+        return _check_state() and _check_perms(module) and _check_type() and _check_size()
+
+    def dump(self):
+        # return result as a dict
+
+        """Serialize the object into a dictionary."""
+
+        result = {
+            'changed': self.changed,
+            'size': self.size,
+            'type': self.type,
+            'filename': self.path,
+            'fingerprint': self.fingerprint,
+        }
+
+        return result
+
+    def remove(self):
+        """Remove the resource from the filesystem."""
+
+        try:
+            os.remove(self.path)
+            self.changed = True
```

Comment:
Shouldn't `changed` only be set to `True` if something is actually deleted? Otherwise, calling the module twice with `state == absent` will result in the module reporting a change twice (i.e. idempotency is violated).

Reply:
This is 1:1 the same code as in https://github.com/ansible/ansible/blob/8a671f54ddc5ec114ef807dc71f49d61261d2107/lib/ansible/module_utils/crypto.py#L150-L160

Reply:
Ah, I think the point is that `os.remove()` raises an exception if the file isn't there, so `changed` won't be set to `True`.